staging/src/k8s.io/api/admissionregistration/v1beta1/types.go

```diff
@@ -576,31 +592,575 @@ type ServiceReference struct {
+	// - expressionIndex: The index of the failed validations in the ValidatingAdmissionPolicy
+	// - validationActions: The enforcement actions enacted for the validation failure
+	// Example audit annotation:
+	// `"validation.policy.admission.k8s.io/validation_failure": "[{\"message\": \"Invalid value\", {\"policy\": \"policy.example.com\", {\"binding\": \"policybinding.example.com\", {\"expressionIndex\": \"1\", {\"validationActions\": [\"Audit\"]}]"`
+	//
+	// Clients should expect to handle additional values by ignoring
+	// any values not recognized.
+	//
+	// "Deny" and "Warn" may not be used together since this combination
+	// needlessly duplicates the validation failure both in the
+	// API response body and the HTTP warning headers.
 	//
 	// Required.
-	Expression string `json:"expression" protobuf:"bytes,2,opt,name=expression"`
+	// +listType=set
+	ValidationActions []ValidationAction `json:"validationActions,omitempty" protobuf:"bytes,4,rep,name=validationActions"`
+}
+
+// BindingParamRef describes how to locate the param used as input to the a
+// expression of rule applied by a policy binding.
+// +union
+type BindingParamRef struct {
+	// `kind` indicates which one of the other fields is non-empty.
+	// Required
+	// +unionDiscriminator
+	Kind BindingParamKind `json:"kind" protobuf:"bytes,1,opt,name=kind"`
+
+	// GlobalParamRef specifies the parameter resource used to configure the
+	// admission control policy.
+	// If the resource referred to by ParamRef does not exist, this binding is
+	// considered mis-configured and the FailurePolicy of the
+	// ValidatingAdmissionPolicy applied.
+	// +optional
+	GlobalParamRef *ParamRef `json:"globalParamRef,omitempty" protobuf:"bytes,2,rep,name=globalParamRef"`
+
+	// NamespaceParamRef specifies the parameter resource or resources used to
+	// configure the admission control policy. It should point to a resources of
+	// the type specified in ParamKind of the bound ValidatingAdmissionPolicy.
+	//
+	// If the policy specifies a ParamKind and the resource referred to by
+	// ParamRef does not exist, this binding is considered mis-configured and
+	// the FailurePolicy of the ValidatingAdmissionPolicy is applied.
+	//
+	// if the paramKind of the policy referred to by `policyName` is cluster scoped,
+	// and namespaceParamRef set, the binding is considered mis-configured, and
+	// the failureMode applies.
+	//
+	// `namespaceParamRef` and `paramRef` are members of a union; if one of the
+	// fields is set, the other must be unset.
+	// +optional
+	NamespaceParamRef *NamespaceParamRef `json:"namespaceParamRef,omitempty" protobuf:"bytes,3,opt,name=namespaceParamRef"`
+}
+
+// BindingParamKind is the kind of ValidatingAdmissionPolicyBinding.
+// +enum
+type BindingParamKind string
+
+// Supported binding param kinds.
+const (
+	BindingParamKindGlobal    BindingParamKind = "Global"
+	BindingParamKindNamespace BindingParamKind = "Namespace"
```

Comment:
`Global` (maybe `ClusterWide`) and `PerNamespace`?  Only calling out the distinction since it's possible to use `ParamRef` to reference a parameter from one namespace and apply it to all namespaces.

Reply:
renamed as suggested